zfs-replicate.sh

```diff
@@ -1,524 +1,538 @@
+    echo $$ > "$lockFile"
+  fi
 }
 
 ## delete lockfiles
 clear_lock() {
-    ## delete lockfiles...and that's all we do here
-    if [ -f "${1}" ]; then
-        printf "Deleting lockfile: %s\n" "${1}"
-        rm "${1}"
-    fi
+  local lockFile=$1
+  ## delete lockfiles...and that's all we do here
+  if [ -f "$lockFile" ]; then
+    printf "Deleting lockfile: %s\n" "$lockFile"
+    rm -f "$lockFile"
+  fi
 }
 
 ## check remote system health
 check_remote() {
-    ## do we have a remote check defined
-    if [ "${REMOTE_CHECK}x" != 'x' ]; then
-        ## run the check
-        $REMOTE_CHECK > /dev/null 2>&1
-        ## exit if above returned non-zero
-        if [ $? != 0 ]; then
-            printf "ERROR: Remote health check '%s' failed!\n" "${REMOTE_CHECK}"
-            exit_error
-        fi
-    fi
+  ## do we have a remote check defined
+  if [ "${REMOTE_CHECK}" == "" ]; then
+    return
+  fi
+  ## run the check
+  if ! $REMOTE_CHECK > /dev/null 2>&1; then
+    exit_clean $? "Remote check '${REMOTE_CHECK}' failed!"
+  fi
 }
 
 ## push replication function
 do_push() {
-    ## check our push lockfile
-    check_lock "${LOGBASE}/.push.lock"
-    ## create initial push command based on arguments
-    ## if first snapname is NULL we do not generate an inremental
-    if [ ${MODE} = PUSH ] && [ ${TYPE} = REMOTE ]; then
-		local dest_snap="$(ssh ${REMOTE_SERVER} $ZFS list -t snapshot -o name | grep ${remote_set} 2>/dev/null | grep autorep- 2>/dev/null | awk -F'@' '{print $2}' | tail -n 1)"
-    elif [ ${MODE} = PUSH ] && [ ${TYPE} = LOCAL ]; then
-		local dest_snap="$(${ZFS} list -t snapshot -o name | grep ${remote_set} 2>/dev/null | grep autorep- 2>/dev/null | awk -F'@' '{print $2}' | tail -n 1)"
-    fi
-    local source_snap="$($ZFS list -t snapshot -o name | grep ${dest_snap} 2>/dev/null | awk -F'@' '{print $2}' | tail -n 1)"
-    local common_snap="${local_set}@${dest_snap}"
-    local receiveargs="-vFd"
-    if [ "${1}" == "NULL" ] && [ ${RECURSE_CHILDREN} -eq 1 ]; then
+  ## check our push lockfile
+  check_lock "${LOGBASE}/.push.lock"
+  ## create initial push command based on arguments
+  ## if first snapname is NULL we do not generate an inremental
+  if [ "$MODE" == "${MODE_PUSH}" ] && [ "$TYPE" == "${TYPE_REMOTE}" ]; then
+    local dest_snap
+    dest_snap="$(ssh "${REMOTE_SERVER}" "$ZFS" list -t snapshot -o name | grep ${remote_set} 2> /dev/null | grep autorep- 2> /dev/null | awk -F'@' '{print $2}' | tail -n 1)"
+  elif [ ${MODE} = PUSH ] && [ ${TYPE} = LOCAL ]; then
+    local dest_snap="$(${ZFS} list -t snapshot -o name | grep ${remote_set} 2> /dev/null | grep autorep- 2> /dev/null | awk -F'@' '{print $2}' | tail -n 1)"
+  fi
+  local source_snap="$($ZFS list -t snapshot -o name | grep ${dest_snap} 2> /dev/null | awk -F'@' '{print $2}' | tail -n 1)"
+  local common_snap="${local_set}@${dest_snap}"
+  local receiveargs="-vFd"
+  if [ "${1}" == "NULL" ] && [ ${RECURSE_CHILDREN} -eq 1 ]; then
```

Comment:
Let's also try to remove the generic $1, $2, etc. in the code and map them to local variables at the top of the function.

Reply:
So, the purpose of the source and common snap was essentially to 

1. Check the destination for the latest snapshot
2. Check if the corresponding snapshot exists on the source
3. If it doesn't, then error out
4. If it does exist, then do an incremental replication from one to the other

Reply:
How would we do this?

$1 = varname

or...?

Reply:
something like ...
```bash
local snapshot=$1
```